JMMServer/Tasks/RecreateAllGroupsTask.cs

```diff
@@ -0,0 +1,373 @@
+            }
+
+            using (ITransaction trans = session.BeginTransaction())
+            {
+                _animeGroupRepo.InsertBatch(session, newGroupsToSeries.Select(gts => gts.Item1));
+                trans.Commit();
+            }
+
+            // Anime groups should have IDs now they've been inserted. Now assign the group ID's to their respective series
+            // (The caller of this method will be responsible for saving the AnimeSeries)
+            foreach (Tuple<AnimeGroup, AnimeSeries> groupAndSeries in newGroupsToSeries)
+            {
+                groupAndSeries.Item2.AnimeGroupID = groupAndSeries.Item1.AnimeGroupID;
+            }
+
+            _log.Info("Generated {0} AnimeGroups", newGroupsToSeries.Length);
+
+            return newGroupsToSeries.Select(gts => gts.Item1);
+        }
+
+        /// <summary>
+        /// Creates <see cref="AnimeGroup"/> that contain <see cref="AnimeSeries"/> that appear to be related.
+        /// </summary>
+        /// <remarks>
+        /// This method assumes that there are no active transactions on the specified <paramref name="session"/>.
+        /// </remarks>
+        /// <param name="session">The NHibernate session.</param>
+        /// <param name="seriesList">The list of <see cref="AnimeSeries"/> to create groups for.</param>
+        /// <returns>A sequence of the created <see cref="AnimeGroup"/>s.</returns>
+        private IEnumerable<AnimeGroup> AutoCreateGroupsWithRelatedSeries(ISessionWrapper session, IReadOnlyCollection<AnimeSeries> seriesList)
+        {
+            _log.Info("Auto-generating AnimeGroups for {0} AnimeSeries based on aniDB relationships", seriesList.Count);
+
+            DateTime now = DateTime.Now;
+            Regex truncateYearRegex = new Regex(@"\s*\(\d{4}\)$");
+            var grpCalculator = AutoAnimeGroupCalculator.Create(session);
+
+            _log.Info("The following exclusions will be applied when generating the groups: " + grpCalculator.Exclusions);
+
+            // Group all of the specified series into their respective groups (keyed by the groups main anime ID)
+            var seriesByGroup = seriesList.ToLookup(s => grpCalculator.GetGroupAnimeId(s.AniDB_ID));
+            var newGroupsToSeries = new List<Tuple<AnimeGroup, IReadOnlyCollection<AnimeSeries>>>(seriesList.Count);
+
+            foreach (var groupAndSeries in seriesByGroup)
+            {
+                int mainAnimeId = groupAndSeries.Key;
+                AnimeSeries mainSeries = groupAndSeries.FirstOrDefault(series => series.AniDB_ID == mainAnimeId);
+                AnimeGroup animeGroup = new AnimeGroup();
+                string groupName = null;
+
+                if (mainSeries != null)
+                {
+                    animeGroup.Populate(mainSeries, now);
+                    groupName = mainSeries.GetSeriesName(session);
+                }
+                else // The anime chosen as the group's main anime doesn't actually have a series

```

Comment:
Good fallback, but if the series is null, that is an issue. AniDB_Anime.CreateAnimeSeriesAndGroup should called on every AniDB_Anime, so lacking an AnimeSeries is a huge problem that will cause an issue elsewhere.